text/0032-dust-protection/0032-dust-protection.md

```diff
@@ -0,0 +1,114 @@
+
+
+### The new protocol dust rule
+
+Let `A` be the address that should hold the dust outputs' balances. Let `M` be the sum of all SigLockedDustAllowanceOutputs' values on `A`.So the total allowed dust outputs on `A` is (`M`/1 MIOTA) * 100 rounded down to the nearest integer. So 100 outputs for 1 MIOTA deposited.
+
+The SigLockedDustAllowanceOutput must be greater or equal to 1MI.
+
+If while processing a milestone a new dust UTXO is created but there is not a sufficient amount already locked in SigLockedDustAllowanceOutputs then the encompassing message should be marked as `ignored`.
+
+If while processing a milestone an unspent SigLockedDustAllowanceOutput is consumed and as a result we have a violation on the number of allowed dust outputs then the encompassing message should be marked as `ignored`.
+
+The checks should happen after the entire message was processed. Messages that were previously applied by the current milestone are taken into account.
+
+
+# Drawbacks
+
+1. We can't have an address that will ever hold less than 1 MI. It will be a bit hard on testnet faucets but still workable.
+2. A service receiving micropayments may fail receiving them if it didn't consolidate dust outputs or raised the deposit for the address.
+
+# Rationale and alternatives
+
+Other solutions that were discussed:
+
+*Burning Dust* - allow dust to live for a limited amount of time on the ledger before removing it from total supply.
+
+*Sweeping dust into merkle tree* - Instead of storing all of the dust in the UTXO set, some nodes can accumulate the dust outputs to a merkle tree root. When the dust is spent, those nodes will require a merkle proof of inclusion.
+
+The first solution has some attack vectors that may burn dust that a user may have wanted to consolidate. Also solutions changing the supply can be quite controversial.
+
+The second solution, besides being more complicated, will relief the problem from one class of nodes but not all nodes.
+
+
+# Unresolved questions
+
+- An attacker may try to fill an address with dust to fill the total cap allowed and block honest senders that want to pay the service. The service can protect by simply consolidating the attacker dust and collecting it for profit. The problem is that if the cost of doing PoW too often may exceed the bad dust profit. Perhaps it is a good idea to limit "tiny" (1 i) dust outputs more than larger dust outputs? In the original discussion each dust class was supposed to be capped to a certain fixed amount. Perhaps we should give weights to different dust classes? So "tiny" dust will weigh more and fill the cap more quickly than larger dust?
+
+- Total cap per address so db key won't be polluted... should this be part of the RFC? This also depends on how key-value dbs will behave at the implementation level. RocksDb for example can configured to handle this well.
+
+- If we stay with the naive linear scheme, is 100 outputs per 1MI is really our choice?
```

Comment:
This is a tricky point, since the deposit/dust_outputs ratio threshold can only be decreased once in place, otherwise we'd invalidate confirmed UTXOs. So we should rather start with a really conservative (high) deposit for dust UTXOs number.

Reply:
Is this really an issue? Since the validity of dust transaction is only evaluated when they become confirmed, it would still work as expected even though the dust state on certain address might technically be invalid.

Reply:
It is true that you are not invalidating the unlock condition, but just the confirmation condition; nonetheless it will undermine validation of the past, in which case you will have to strictly rely on proofs of inclusions, without being able to make up your own mind on what is valid or not.

Reply:
> This is a tricky point, since the deposit/dust_outputs ratio threshold can only be decreased once in place
The ratio can't be decreased below the validation constraints